atom/renderer/atom_autofill_agent.h

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2017 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#ifndef ATOM_RENDERER_ATOM_AUTOFILL_AGENT_H_
+#define ATOM_RENDERER_ATOM_AUTOFILL_AGENT_H_
+
+#include <vector>
+
+#include "base/memory/weak_ptr.h"
+#include "content/public/renderer/render_frame_observer.h"
+#include "content/public/renderer/render_view_observer.h"
+#include "third_party/WebKit/public/web/WebAutofillClient.h"
+#include "third_party/WebKit/public/web/WebFormControlElement.h"
+#include "third_party/WebKit/public/web/WebInputElement.h"
+#include "third_party/WebKit/public/web/WebNode.h"
+
+namespace atom {
+
+class AutofillAgent : public content::RenderFrameObserver,
+                                public blink::WebAutofillClient {
+ public:
+  explicit AutofillAgent(content::RenderFrame* frame);
+  ~AutofillAgent();
+
+  // content::RenderFrameObserver:
+  void OnDestruct() override;
+
+  void DidChangeScrollOffset() override;
+  void FocusedNodeChanged(const blink::WebNode&) override;
+
+ private:
+  class Helper : public content::RenderViewObserver {
+   public:
+    explicit Helper(AutofillAgent* agent);
+
+    // content::RenderViewObserver implementation.
+    void OnDestruct() override {}
+    void OnMouseDown(const blink::WebNode&) override;
+    void FocusChangeComplete() override;
+
+   private:
+    AutofillAgent* agent_;
+  };
+  friend class Helper;
+
+  struct ShowSuggestionsOptions {
+    ShowSuggestionsOptions();
+    bool autofill_on_empty_values;
+    bool requires_caret_at_end;
+  };
+
+  bool OnMessageReceived(const IPC::Message& message) override;
+  void OnWebContentsRoutingIdReceived(int routing_id);
+
+  // blink::WebAutofillClient:
+  void textFieldDidEndEditing(const blink::WebInputElement&) override;
+  void textFieldDidChange(const blink::WebFormControlElement&) override;
+  void textFieldDidChangeImpl(const blink::WebFormControlElement&);
+  void textFieldDidReceiveKeyDown(const blink::WebInputElement&,
+                                  const blink::WebKeyboardEvent&) override;
+  void openTextDataListChooser(const blink::WebInputElement&) override;
+  void dataListOptionsChanged(const blink::WebInputElement&) override;
+
+  bool IsUserGesture() const;
+  void HidePopup();
+  void ShowPopup(const blink::WebFormControlElement&,
+                 const std::vector<base::string16>&,
+                 const std::vector<base::string16>&);
+  void ShowSuggestions(const blink::WebFormControlElement& element,
+                       const ShowSuggestionsOptions& options);
+  void OnAcceptSuggestion(base::string16 suggestion);
+
+  void DoFocusChangeComplete();
+
+  content::RenderFrame* render_frame_;
+  int web_contents_routing_id_;
```

Comment:
There is no need to record `render_frame_` nor `web_contents_routing_id_`, the `RenderFrameObserver` already provides helpers to get them.